src/adm.rs

```diff
@@ -1,13 +1,79 @@
 use serde::{Deserialize, Serialize};
+use std::collections::{BTreeSet, HashMap};
 use url::Url;
 
 use crate::error::HandlerError;
+use crate::settings::Settings;
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct AdmTileResponse {
     pub tiles: Vec<AdmTile>,
 }
 
+/// Filter criteria for adm Tiles
+#[derive(Default, Clone, Debug)]
+pub struct AdmFilter {
+    /// list of allowed base host strings.
+    pub allowed_url_hosts: Option<BTreeSet<String>>,
+    pub field_defaults: HashMap<String, String>,
+}
+
+impl AdmFilter {
+    /// Filter and process tiles from ADM:
+    ///
+    /// - Returns None for tiles that shouldn't be shown to the client
+    /// - Modifies tiles for output to the client (adding additional fields, etc.)
+    pub fn filter_and_process(&self, tile: AdmTile) -> Option<AdmTile> {
+        let host: Url = match tile.advertiser_url.parse() {
+            Ok(v) => v,
+            Err(e) => {
+                warn!(
+                    "Could not parse advertiser URL {:?} : {:?}",
+                    tile.advertiser_url, e
+                );
+                return None;
+            }
+        };
+        // Use strict matching for now, eventually, we may want to use backwards expanding domain
+        // searches, (.e.g "xyz.example.com" would match "example.com")
+        match self.allowed_url_hosts.clone() {
+            Some(allowed) => {
+                let host_name = host.host_str().unwrap_or("UKNOWN");
+                if allowed.contains(host_name) {
+                    dbg!("👍🏻", host_name);
+                    return Some(tile);
+                }
+                dbg!("👎🏻", host_name);
+                None
+            }
+            None => Some(tile),
+        }
+    }
+}
+
+/// Construct the AdmFilter from the provided settings.
+/// This uses `allowed_vendors` (a JSON formatted list of strings),
+///
+impl From<&Settings> for AdmFilter {
+    fn from(settings: &Settings) -> Self {
+        let mut allowed_url_hosts: BTreeSet<String> = BTreeSet::new();
+        if let Some(hosts) = settings.clone().allowed_vendors {
+            for host in hosts {
+                allowed_url_hosts.insert(host);
+            }
+        };
+
+        AdmFilter {
+            allowed_url_hosts: if allowed_url_hosts.is_empty() {
+                None
```

Comment:
I'd rather this not be an `Option` but it could be useful to have an `allow all` for local debugging (that acts like the current `None` behavior here)

Reply:
I kept debating about the `Option` as well, but it struck me that it could be possible for ADM filter to return a valid response, that contained no valid entries. If that's not really possible, then I'm fine with this returning a `Result<AdmFilter>`